CodeEditModules/Modules/WorkspaceClient/src/Live.swift

```diff
@@ -54,41 +58,141 @@ public extension WorkspaceClient {
+
+            // get currently indexed children of directory
+            var newChildren = fileItem.children?.filter({ _ in true }) // build a new array based on fileItem.children
+
+            // test for deleted children, and remove them from the index
+            for oldContent in fileItem.children ?? [] {
+                if directoryContentsUrls.contains(oldContent.url) {
+                    continue
+                }
+
+                if let removeAt = newChildren?.firstIndex(of: oldContent) {
+                    newChildren?.remove(at: removeAt)
+                    flattenedFileItems.removeValue(forKey: oldContent.id)
+                }
+            }
+
+            // test for new children, and index them using loadFiles
+            for newContent in directoryContentsUrls {
+                guard !ignoredFilesAndFolders.contains(newContent.lastPathComponent) else { continue }
+
+                var childExists = false
+                fileItem.children?.forEach({ childExists = $0.url == newContent ? true : childExists })
+                if childExists {
+                    continue
+                }
+
+                var isDir: ObjCBool = false
+                if fileManager.fileExists(atPath: newContent.path, isDirectory: &isDir) {
+                    var subItems: [FileItem]?
+
+                    if isDir.boolValue {
+                        subItems = try loadFiles(fromURL: newContent)
+                    }
+
+                    let newFileItem = FileItem(url: newContent, children: subItems?.sortItems(foldersOnTop: true))
+                    subItems?.forEach {
+                        $0.parent = newFileItem
+                    }
+                    newFileItem.parent = fileItem
+                    flattenedFileItems[newFileItem.id] = newFileItem
+                    newChildren?.append(newFileItem)
                 }
             }
 
-            source?.setCancelHandler {
-                close(descriptor)
+            newChildren?.forEach({
+                try? rebuildFiles(fromItem: $0)
+                flattenedFileItems[$0.id] = $0
+            })
+
+            fileItem.children = newChildren
+            return
+        }
+
+        var sources: [String: DispatchSourceFileSystemObject] = [:]
+
+        /// Function to apply listeners that rebuild the file index when the file system is changed.
+        /// Optimised so that it only deletes/creates the needed listeners instead of replacing every one.
+        func startListeningToDirectory() {
+            // iterate over every item, checking if its a directory first
+            for item in flattenedFileItems.values {
+                // check if it actually exists, doesn't have a listener, and is a folder
+                guard item.isFolder else { continue }
+                guard !sources.keys.contains(item.id) else { continue }
+                guard FileItem.fileManger.fileExists(atPath: item.url.path) else { continue }
```

Comment:
Since you are not throwing any errors, i think you can better chain this guard.

```swift
guard item.isFolde,
           !sources.keys.contains(item.id)
           FileItem.fileManger.fileExists(atPath: item.url.path) else { continue }
```